src/expr/src/scalar/mod.rs

```diff
@@ -884,6 +884,44 @@ impl MirScalarExpr {
                                     })
                                     .collect(),
                             }
+                        } else if let (
+                            BinaryFunc::Eq,
+                            literal @ MirScalarExpr::Literal(Ok(..), ..),
+                            MirScalarExpr::CallUnary { func, expr },
+                        ) = (&*func, &**expr1, &**expr2)
+                        {
+                            // <lit> = f(<expr>), where f is invertible
+                            //  -->
+                            // <f^-1(lit)> = <expr>
+                            // (if f^-1(lit) doesn't error out)
+                            //
+                            // This enables index usage when an implicit cast was inserted on a
+                            // column reference. E.g., if `a` is smallint, then `a = 5` is
+                            // `smallint_to_integer(a) = 5`, which we now transform to `a = 5`.
+                            if let Some(inverse_func) = func.invert() {
+                                let inved_lit = eval(&MirScalarExpr::CallUnary {
+                                    func: inverse_func,
+                                    expr: Box::new(literal.clone()),
+                                });
+                                // The evaluation can error out, e.g., when casting a too large
+                                // int32 to int16. In such a case, we currently avoid doing any
+                                // transformation here. This will unfortunately prevent index usage.
+                                // TODO: We should do something smart when the eval errors out.
+                                // There are several pitfalls to avoid:
+                                // - Rewriting the literal to a literal error is incorrect: e.g.
+                                //   assume `a` is smallint, `a=3000000000 or a=7`. Here, the entire
+                                //   expression would error out, while the original expression is
+                                //   true when `a=7` is true.
+                                // - Rewriting *e to `false` is incorrect: e.g.,
+                                //   `(a = 3000000000) IS NULL`, should be `true` when `a` is null.
+                                if !inved_lit.is_literal_err() {
+                                    *e = MirScalarExpr::CallBinary {
+                                        func: BinaryFunc::Eq,
+                                        expr1: Box::new(inved_lit),
+                                        expr2: expr.clone(),
```

Comment:
Same as above: no need to `clone` as the `expr` is going to be dropped at the end of this block, so you can just take ownership.

Reply:
Done